### ISSUE TITLE
New rule option at Eventchannel to filter dynamically by the same id, user, source IP, etc.

### ISSUE THREAD

**Issue by cristgl** [wazuh/wazuh#2531]
|Wazuh version|Install type|Install method|Platform|
|---|---|---|---|
| 3.8.2 | Manager/Agent | Packages/Sources | Windows |

At the time, there is no possibility to filter by the same data at the new Eventchannel, as rule options like `<same_user />`, `<same_source_ip />', etc. don't match. This is due to the `user` static variable. This variable only exists if the event is decoded with the Windows decoder, which is not happening with the new Eventchannel. To accomplish these rule options in the new log format, we propose one only option, where it would be possible to set the field dynamically, as stated below:

```
  <rule id="20086" level="10" frequency="$MS_FREQ" timeframe="240">
    <if_matched_group>win_authentication_failed</if_matched_group>
    <same_field>EventChannel.EventData.TargetUserName</same_field>
    <description>Multiple Windows Logon Failures</description>
    <options>no_full_log</options>
    <group>authentication_failures,pci_dss_10.2.4,pci_dss_10.2.5,pci_dss_11.4,gdpr_IV_35.7.d,gdpr_IV_32.2,</group>
  </rule>
```

**Comment by chemamartinez** [comment]
Working on branch [same-dynamic-fields](https://github.com/wazuh/wazuh/tree/same-dynamic-fields).

Currently, it adds two new options to rules, `same_field` and `not_same_field` to be able to compare a dynamic field from an event with the previous one. This way we can correlate alerts.

Right now, it only works for a tag in each rule. It can be improved by adding support for more than one field at the same time.

Here we can see a simple example of how it works. To do that, we create the following custom rules:

``` xml
<rule id="100001" level="3">
    <if_sid>221</if_sid>
    <field name="netinfo.iface.name">ens33</field>
    <options>no_full_log</options>
    <description>ens33 interface enabled</description>
  </rule>

  <rule id="100002" level="7" frequency="3" timeframe="300">
    <if_matched_sid>100001</if_matched_sid>
    <same_field>netinfo.iface.name</same_field>
    <options>no_full_log</options>
    <description>Repeated iface name ens33 three times</description>
  </rule>
```

After that, we wait for three executions of the inventory scan, where the interface `ens33` won't change its name. So we get the following alerts:

```
** Alert 1550852339.328327: - local,syscollector,
2019 Feb 22 08:18:59 ubuntu->syscollector
Rule: 100001 (level 3) -> 'ens33 interface enabled'
type: network
netinfo.iface.name: ens33
netinfo.iface.type: ethernet
netinfo.iface.state: up
netinfo.iface.mtu: 1500
netinfo.iface.mac: 00:0c:29:58:1c:4c
netinfo.iface.tx_packets: 7110098
netinfo.iface.rx_packets: 7491740
netinfo.iface.tx_bytes: 948617545
netinfo.iface.rx_bytes: 1095600003
netinfo.iface.tx_errors: 0
netinfo.iface.rx_errors: 0
netinfo.iface.tx_dropped: 0
netinfo.iface.rx_dropped: 0
netinfo.iface.ipv4.gateway: unknown
netinfo.iface.ipv4.dhcp: disabled
netinfo.iface.ipv4.address: 172.16.98.128
netinfo.iface.ipv4.netmask: 255.255.255.0
netinfo.iface.ipv4.broadcast: 172.16.98.255
netinfo.iface.ipv6.dhcp: enabled
netinfo.iface.ipv6.address: fe80::20c:29ff:fe58:1c4c
netinfo.iface.ipv6.netmask: ffff:ffff:ffff:ffff::
```

At the third scan, we get the following alert instead:

```
** Alert 1550852369.329745: - local,syscollector,
2019 Feb 22 08:19:29 ubuntu->syscollector
Rule: 100002 (level 7) -> 'Repeated iface name ens33 three times'
type: network
netinfo.iface.name: ens33
netinfo.iface.type: ethernet
netinfo.iface.state: up
netinfo.iface.mtu: 1500
netinfo.iface.mac: 00:0c:29:58:1c:4c
netinfo.iface.tx_packets: 7110246
netinfo.iface.rx_packets: 7491918
netinfo.iface.tx_bytes: 948725365
netinfo.iface.rx_bytes: 1095614424
netinfo.iface.tx_errors: 0
netinfo.iface.rx_errors: 0
netinfo.iface.tx_dropped: 0
netinfo.iface.rx_dropped: 0
netinfo.iface.ipv4.gateway: unknown
netinfo.iface.ipv4.dhcp: disabled
netinfo.iface.ipv4.address: 172.16.98.128
netinfo.iface.ipv4.netmask: 255.255.255.0
netinfo.iface.ipv4.broadcast: 172.16.98.255
netinfo.iface.ipv6.dhcp: enabled
netinfo.iface.ipv6.address: fe80::20c:29ff:fe58:1c4c
netinfo.iface.ipv6.netmask: ffff:ffff:ffff:ffff::
```

**Comment by branchnetconsulting** [comment]
Awesome, Chema!  I've been looking forward to this.  I love the <same_field>, <not_same_field> approach and the way it completely replaces the clunky old static versions of the same.  Before we know it there won't be any more of those old tags dedicated to the legacy static fields from OSSEC anywhere to be found in Wazuh!   That will be a day to celebrate :)

**Comment by chemamartinez** [comment]
Hi @branchnetconsulting ,

Yeah, I am happy you like it.

I agree with you it will be very useful for the ruleset and the analysis engine.
The PR has been already created (#2689) and it will be available in the next released version.

For future versions, we also want to deprecate options like `same_srcip` in favor of `<same_field>srcip</same_field>`.

Regards,
Chema.

**Comment by chemamartinez** [comment]
Merged: #2689
